### PR TITLE
License rework

### PR DESCRIPTION
**Pull request by rmfranken**
Good lord why does it seem impossible to solve this merge conflict. Just trying to draft this pull request here to see if it's resolvable in the Github Web UI because pycharm is causing me to have a stroke.

### CODE REVIEW

**Comment by rmfranken** [comment]
Something is up. I just tried pushing a function rename commit from my local to this branch - and it says it managed - except that commit is nowhere to be found here...
image

**Comment by rmfranken** [comment]
Oh NOW it shows up of course. Ok. I'm going insane ahahah

File: gimie/sources/common/license.py [hunk 1]
```diff
@@ -1,56 +1,125 @@
...
+        return "More than 1 license file was found, please make sure you only have one license."
+    else:
+        for license_file_url in license_files:
+            return license_file_url
+
 
-        # The first root of os.walk is the current dir
-        if not recurse:
-            return license_files
+def github_read_file(url: str, github_token: str = None) -> dict:
```
Comment (cmdoret) [inline gimie/sources/common/license.py:88]: If above is implemented,this function will not be needed here (defined as a method of the file object)

Comment (cmdoret) [reply gimie/sources/common/license.py:88]: headers will then be passed down from the Extractor to the File

File: gimie/sources/common/license.py [hunk 2]
```diff
@@ -1,56 +1,125 @@
...
+        "detected_license_expression_spdx"
+    ]
+    return found_spdx_license_id
 
 
 def get_spdx_url(name: str) -> str:
-    """Given an SPDX license identifier, return the full URL."""
-    return f"https://spdx.org/licenses/{name}"
+    """Given an SPDX license identifier, return the full SPDX URL."""
+    return f"https://spdx.org/licenses/{name}.html"
```
Comment (cmdoret) [inline gimie/sources/common/license.py:115]: might be overkill to have it as a function if used only once

File: pyproject.toml [hunk 3]
```diff
@@ -35,6 +35,9 @@ python-dotenv = "^0.21.1"
 pre-commit = "^3.0.0"
 importlib = "^1.0.4"
 python-dateutil = "^2.8.2"
+spdx-lookup = "^0.3.3"
```
Comment (cmdoret) [inline pyproject.toml:38]: do we still need spdx-lookup and spdx-matcher?

File: gimie/sources/github.py [hunk 4]
```diff
@@ -129,10 +129,11 @@ def extract(self):
...
         self.date_created = isoparse(data["createdAt"][:-1])
         self.date_modified = isoparse(data["updatedAt"][:-1])
+        self.license = self._get_license()
         # If license is available, convert to standard SPDX URL
-        if data["licenseInfo"] is not None:
-            if "NOASSERTION" not in str(data["licenseInfo"]):
-                self.license = get_spdx_url(data["licenseInfo"]["spdxId"])
+        # if data["licenseInfo"] is not None:
+        #     if "NOASSERTION" not in str(data["licenseInfo"]):
+        #         self.license = get_spdx_url(data["licenseInfo"]["spdxId"])
```
Comment (cmdoret) [inline gimie/sources/github.py:136]:
```suggestion

```
I think these could be removed.

File: gimie/sources/github.py [hunk 5]
```diff
@@ -284,6 +285,11 @@ def _get_user(self, node: Dict[str, Any]) -> Person:
             affiliations=orgs,
         )
 
+    def _get_license(self):
+        """Extract a SPDX License URL from a GitHub Repository"""
+        self.license = get_license(repo_url=self.url, headers=self._set_auth())
```
Comment (cmdoret) [inline gimie/sources/github.py:290]: In the future, `get_license` should take a list of `File` objects. This would allow it to work with Git or Github.

**Review by cmdoret — CHANGES_REQUESTED** [review]
Nice, the logic is there and indeed we need to integrate it with the Extractor classes.
Thanks for adding the docstrings and type annotations! I'll work on a separate PR to add a generic `File` object that you could use in your license functions.

**Comment by rmfranken** [comment]
Ok, @cmdoret I think I'm ready for your review, only 65 commits later. 🛠️ Hopefully it's the last one!

File: gimie/sources/common/license.py [hunk 6]
```diff
@@ -1,56 +1,34 @@
...
-    return license_files
-
-
-def get_spdx_url(name: str) -> str:
-    """Given an SPDX license identifier, return the full URL."""
-    return f"https://spdx.org/licenses/{name}"
+from gimie.io import Resource, iterable_to_stream, RemoteResource
+
+
+def is_license_path(file: Resource) -> bool:
```
[the remark this answers is not in the document]
Comment (rmfranken) [reply gimie/sources/common/license.py:5]: Understood - that makes more sense indeed.

File: gimie/sources/github.py [hunk 7]
```diff
@@ -16,16 +16,17 @@
...
 from dateutil.parser import isoparse
 from functools import cached_property
 import os
 import requests
-from typing import Any, Dict, List, Optional, Set, Union
+from typing import Any, Dict, List, Optional, Union
 from urllib.parse import urlparse
 from dotenv import load_dotenv
-
+from scancode.api import get_licenses
```
[the remark this answers is not in the document]
Comment (rmfranken) [reply gimie/sources/github.py:29]: Love the term shotgun surgery.

that being said, I see your reasoning - and can get behind it, even with the heavy price of defining dummy functions which I would argue reduce readability.

File: gimie/sources/gitlab.py [hunk 8]
```diff
@@ -263,6 +283,27 @@ def _get_user(self, node: Dict[str, Any]) -> Person:
...
+    def _get_license(self) -> list[str]:
+        """Extract a SPDX License URL from a GitLab Repository"""
+
+        license_files_iterator = filter(is_license_path, self.list_files())
+        license_files = list(license_files_iterator)
+        license_ids = []
+        for file in license_files:
+            with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+                temp_file.write(file.open().read())
+                license_detections = get_licenses(temp_file.name)[
```
[the remark this answers is not in the document]
Comment (rmfranken) [reply gimie/sources/gitlab.py:295]: This was the spiciest one, but I think this works. I was messing around a bit much with the for loop + filling the list of license_id's so I left that in the github+gitlab part, but now the fetching of the license detections and the highest coverage function happen in the generic license.py.

**Pull request merged**